src/Function/Metric/Bundles.agda

```diff
@@ -0,0 +1,139 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Bundles for metrics
+------------------------------------------------------------------------
+
+-- The contents of this module should be accessed via `Function.Metric`.
+
+{-# OPTIONS --without-K --safe #-}
+
+module Function.Metric.Bundles  where
+
+open import Algebra.Core using (Op₂)
+open import Level using (Level; suc; _⊔_)
+open import Relation.Binary.Core using (Rel)
+
+open import Function.Metric.Structures
+open import Function.Metric.Core
+
+------------------------------------------------------------------------
+-- ProtoMetric
+
+record ProtoMetric (a i ℓ₁ ℓ₂ ℓ₃ : Level)
+                 : Set (suc (a ⊔ i ⊔ ℓ₁ ⊔ ℓ₂ ⊔ ℓ₃)) where
+  field
+    Carrier       : Set a
+    Image         : Set i
+    _≈_           : Rel Carrier ℓ₁
+    _≈ᵢ_          : Rel Image ℓ₂
+    _≤_           : Rel Image ℓ₃
+    0#            : Image
+    d             : DistanceFunction Carrier Image
+    isProtoMetric : IsProtoMetric _≈_ _≈ᵢ_ _≤_ 0# d
+
+  open IsProtoMetric isProtoMetric public
+
+------------------------------------------------------------------------
+-- PreMetric
+
+record PreMetric (a i ℓ₁ ℓ₂ ℓ₃ : Level)
+               : Set (suc (a ⊔ i ⊔ ℓ₁ ⊔ ℓ₂ ⊔ ℓ₃)) where
+  field
+    Carrier     : Set a
+    Image       : Set i
+    _≈_         : Rel Carrier ℓ₁
+    _≈ᵢ_        : Rel Image ℓ₂
+    _≤_         : Rel Image ℓ₃
+    0#          : Image
+    d           : DistanceFunction Carrier Image
+    isPreMetric : IsPreMetric _≈_ _≈ᵢ_ _≤_ 0# d
+
+  open IsPreMetric isPreMetric public
+
+  protoMetric : ProtoMetric a i ℓ₁ ℓ₂ ℓ₃
+  protoMetric = record
+    { isProtoMetric = isProtoMetric
+    }
+
+------------------------------------------------------------------------
+-- QuasiSemiMetric
+
+record QuasiSemiMetric (a i ℓ₁ ℓ₂ ℓ₃ : Level)
+                     : Set (suc (a ⊔ i ⊔ ℓ₁ ⊔ ℓ₂ ⊔ ℓ₃)) where
+  field
+    Carrier           : Set a
+    Image             : Set i
+    _≈_               : Rel Carrier ℓ₁
+    _≈ᵢ_              : Rel Image ℓ₂
+    _≤_               : Rel Image ℓ₃
+    0#                : Image
+    d                 : DistanceFunction Carrier Image
+    isQuasiSemiMetric : IsQuasiSemiMetric _≈_ _≈ᵢ_ _≤_ 0# d
+
+  open IsQuasiSemiMetric isQuasiSemiMetric public
+
+  preMetric : PreMetric a i ℓ₁ ℓ₂ ℓ₃
+  preMetric = record
```

Comment:
would it be ok to lay all these out on a single line instead of 3? I completely get that once you get 2 (or more) fields, it does make sense to use more vertical space -- but just 1?

Reply:
I've got into the habit of putting them on 3 because in `Relation.Binary`/`Algebra` hierarchies the names are very long and they don't always fit on one line under the very loosely obeyed 80 characters per line limit. I guess it doesn't matter here?